src/Storages/MergeTree/MergeTreeData.h

```diff
@@ -1378,7 +1378,7 @@ class MergeTreeData : public IStorage, public WithMutableContext
     /// Used to receive AlterConversions for part and apply them on fly. This
     /// method has different implementations for replicated and non replicated
     /// MergeTree because they store mutations in different way.
-    virtual std::map<int64_t, MutationCommands> getAlterMutationCommandsForPart(const DataPartPtr & part) const = 0;
+    virtual std::vector<MutationCommands> getAlterMutationCommandsForPart(const DataPartPtr & part) const = 0;
```

Comment:
Consider adding a comment about whether the returned vector is sorted. Looks like it isn't, and the caller doesn't care about the order? Also mention that only RENAME_COLUMN are included.

(I'm confused by this. I would've expected the order to matter: if a column was renamed twice A -> B -> C then it wouldn't work if we tried to do B -> C before A -> B. But I see that AlterConversions doesn't handle that. I guess such chain-rename is just not handled correctly here? It's also weird that `getAlterMutationCommandsForPart()` doesn't take a metadata snapshot, and instead returns commands to promote the part all the way to the latest schema; I guess that's also incorrect? If it used a metadata snapshot, maybe the whole thing would be correct because at most one ALTER can be in progress at a time, and dependent renames are not allowed within one ALTER. This is all just speculation, I'm not familiar with this code.)

Reply:
> Also mention that only RENAME_COLUMN are included.

Ok, I've added the comment.

> I would've expected the order to matter: if a column was renamed twice A -> B -> C then it wouldn't work if we tried to do B -> C before A -> B. But I see that AlterConversions doesn't handle that.

It don't need to, since only one ALTER_METADATA can be executed concurrently, so if you have two ALTER that modifies metadata the first one should be finished completely, only after that the second can proceed - https://github.com/ClickHouse/ClickHouse/blob/b22118a2dd2ffae0d84c9f52640d24fb8253ede9/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1455.

> It's also weird that getAlterMutationCommandsForPart() doesn't take a metadata snapshot, and instead returns commands to promote the part all the way to the latest schema; I guess that's also incorrect? If it used a metadata snapshot, maybe the whole thing would be correct because at most one ALTER can be in progress at a time, and dependent renames are not allowed within one ALTER

This code is required to handle RENAME COLUMN that had not been finished yet, while the metadata had been already updated, so when query contains new names it can transform new names to old ones, to find this data on the disk. Metadata snapshot does not have information about mutations so it cannot be used here.

Reply:
Does anything prevent the following scenario?:
 1. A SELECT calls `StorageMergeTree::read()`.
 2. `StorageMergeTree::read()` grabs the current list of active parts. Say, it has one part: all_1_1_1, with one column A.
 3. `ALTER RENAME COLUMN A TO B` starts and finishes.
 4. `ALTER RENAME COLUMN B TO C` starts and finishes.
 5. The SELECT from step 1 calls getAlterMutationCommandsForPart() for part all_1_1_1.
 6. getAlterMutationCommandsForPart() returns both renames: A->B, B->C.
 7. AlterConversions gets confused (with or without this PR), and the SELECT fails.
 8. Even if AlterConversions could handle A->B->C correctly, this SELECT would still fail because it expects data to match the schema from the metadata snapshot obtained in step 1, so column C would be unexpected.

Reply:
Non-mutated part (`all_1_1_1`) should not have any mutations (i.e. `getAlterConversionsForPart` should return nothing), so it should be OK.

Reply:
`getAlterConversionsForPart()` -> `getAlterMutationCommandsForPart()` take all mutations that happened in the table (`current_mutations_by_version`) after this part was created (`part_data_version`). For all_1_1_1 that would be both of the renames. On SELECT path, `getAlterConversionsForPart()` is called from `MergeTreeData::getStorageSnapshot()`.

Reproduced this scenario by adding a sleep in the middle of `MergeTreeData::getStorageSnapshot()` and doing ALTER while SELECT is stuck. The SELECT fails even with one ALTER, presumably because the metadata snapshot expects the old column name, while AlterConversions produce the new name:
```
:) desc (select * from a); select * from a;

DESCRIBE TABLE
(
    SELECT *
    FROM a
)

Query id: 0af8ba13-838c-4a21-9337-c0a1c2800636

┌─name─┬─type──┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ k    │ Int64 │              │                    │         │                  │                │
│ x    │ Int64 │              │                    │         │                  │                │
└──────┴───────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

2 rows in set. Elapsed: 0.001 sec. 


SELECT *
FROM a

Query id: b42a9aca-0d37-43bc-92f8-2c001efd203f


Elapsed: 10.014 sec. 

Received exception from server (version 24.2.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Missing columns: 'z' while processing query: 'SELECT k, z FROM a', required columns: 'k' 'z', maybe you meant: 'k' or 'z'. (UNKNOWN_IDENTIFIER)
```
(Also directly confirmed that the `getAlterMutationCommandsForPart()` returns 2 commands in this case, if there were 2 ALTERs.)

So I guess my understanding was correct.

This is all mostly unrelated to this PR, I just wanted to figure out if this code is as broken as it looks.

Reply:
Oh, yes, you are right, actually my comment was wrong it should be the opposite all_1_1_1 should have mutations, so yes, the SELECT will fail.

I guess this can be fixed by looking at which metadata version had bee used for the SELECT, but this is completely different story...